File: packages/python/muninn/llamaindex.py
"""
LlamaIndex integration for Muninn memory system.

Provides memory classes for LlamaIndex agents to persist
conversation history and knowledge across sessions.
"""

from typing import Any, Dict, List, Optional, Sequence
from llama_index.core.memory import BaseMemory
from llama_index.core.llms import ChatMessage, MessageRole
from .client import MuninnClient


class MuninnChatMemory(BaseMemory):
    """
    LlamaIndex chat memory backed by Muninn.
    
    Usage:
        from muninn.llamaindex import MuninnChatMemory
        from llama_index.core.agent import AgentRunner
        
        memory = MuninnChatMemory(
            api_key="muninn_xxx",
            organization_id="my-agent"
        )
        
        agent = AgentRunner.from_llm(
            llm=llm,
            memory=memory
        )
    """
    
    def __init__(
        self,
        api_key: str,
        organization_id: str = "default",
        base_url: str = "https://api.muninn.au",
        token_limit: int = 3000,
        **kwargs
    ):
        """Initialize Muninn chat memory for LlamaIndex."""
        super().__init__(**kwargs)
        self._client = MuninnClient(
            api_key=api_key,
            organization_id=organization_id,
            base_url=base_url
        )
        self._token_limit = token_limit
    
    @classmethod
    def class_name(cls) -> str:
        return "MuninnChatMemory"
    
    def get(self, input: Optional[str] = None, **kwargs) -> List[ChatMessage]:
        """Get chat history from Muninn."""
        query = input or ""
        
        # Search for relevant memories
        memories = self._client.search(
            query=query,
            limit=20
        )
        
        # Convert to ChatMessage format
        messages = []
        for memory in memories:
            content = memory.get("content", "")
            metadata = memory.get("metadata", {})
            role = metadata.get("role", "user")
            
            msg_role = MessageRole.USER if role == "user" else MessageRole.ASSISTANT
            messages.append(ChatMessage(
                role=msg_role,
                content=content,
                additional_kwargs={"memory_id": memory.get("id")}
            ))
        
        return messages
    
    def get_all(self) -> List[ChatMessage]:
        """Get all chat history."""
        return self.get(input="")
    
    def put(self, message: ChatMessage) -> None:
        """Store a message in Muninn."""
        role = "user" if message.role == MessageRole.USER else "assistant"
        
        self._client.store(
            content=message.content,
            memory_type="conversational",
            metadata={
                "role": role,
                "source": "llamaindex"
            }
        )
    
    def set(self, messages: Sequence[ChatMessage]) -> None:
        """Replace all messages (not recommended for Muninn)."""
        # Clear and re-add (Muninn doesn't support bulk replace)
        for message in messages:
            self.put(message)
    
    def reset(self) -> None:
        """Reset memory (not implemented for safety)."""
        pass


class MuninnVectorMemory:
    """
    Vector-based memory for LlamaIndex using Muninn.
    
    Provides semantic search over stored memories.
    Best for RAG applications where you need to find
    relevant context from past conversations.
    
    Usage:
        from muninn.llamaindex import MuninnVectorMemory
        
        memory = MuninnVectorMemory(
            api_key="muninn_xxx",
            organization_id="my-agent"
        )
        
        # Retrieve relevant context
        context = memory.retrieve("user preferences about React")
    """
    
    def __init__(
        self,
        api_key: str,
        organization_id: str = "default",
        base_url: str = "https://api.muninn.au",
        similarity_threshold: float = 0.7,
        **kwargs
    ):
        """Initialize vector memory."""
        self._client = MuninnClient(
            api_key=api_key,
            organization_id=organization_id,
            base_url=base_url
        )
        self._similarity_threshold = similarity_threshold
    
    def add(self, content: str, metadata: Optional[Dict[str, Any]] = None) -> str:
        """Add content to memory."""
        result = self._client.store(
            content=content,
            memory_type="semantic",
            metadata=metadata or {}
        )
        return result.get("id", "")
    
    def retrieve(
        self,
        query: str,
        limit: int = 10,
        **kwargs
    ) -> List[Dict[str, Any]]:
        """Retrieve semantically similar memories."""
        results = self._client.search(
            query=query,
            limit=limit,
            search_type="hybrid"  # Uses both keyword and vector search
        )
        
        # Filter by similarity threshold
        filtered = []
        for result in results:
            similarity = result.get("score", 1.0)
            if similarity >= self._similarity_threshold:
                filtered.append({
                    "content": result.get("content"),
                    "score": similarity,
                    "metadata": result.get("metadata", {}),
                    "id": result.get("id")
                })
        
        return filtered
    
    def delete(self, memory_id: str) -> bool:
        """Delete a memory by ID."""
        try:
            self._client.delete(memory_id)
            return True
        except Exception:
            return False
    
    def clear(self) -> None:
        """Clear all memories (not implemented for safety)."""
        pass


class MuninnKnowledgeGraphMemory:
    """
    Knowledge graph memory for LlamaIndex.
    
    Stores and retrieves facts as entity-relationship triples.
    Best for agents that need structured knowledge about
    entities and their relationships.
    
    Usage:
        from muninn.llamaindex import MuninnKnowledgeGraphMemory
        
        memory = MuninnKnowledgeGraphMemory(
            api_key="muninn_xxx"
        )
        
        # Query entity relationships
        facts = memory.get_entity_facts("James")
        # Returns: [{"predicate": "works_at", "object": "TechCorp"}, ...]
    """
    
    def __init__(
        self,
        api_key: str,
        organization_id: str = "default",
        base_url: str = "https://api.muninn.au",
        **kwargs
    ):
        """Initialize knowledge graph memory."""
        self._client = MuninnClient(
            api_key=api_key,
            organization_id=organization_id,
            base_url=base_url
        )
    
    def get_entity_facts(
        self,
        entity_name: str,
        limit: int = 50
    ) -> List[Dict[str, Any]]:
        """Get all facts about an entity."""
        try:
            facts = self._client.get_entity_facts(entity_name)
            return facts[:limit]
        except Exception:
            return []
    
    def query_relation(
        self,
        subject: str,
        predicate: str
    ) -> List[Dict[str, Any]]:
        """Query for specific subject-predicate combinations."""
        try:
            results = self._client.search_facts(
                query=f"{subject} {predicate}",
                limit=10
            )
            return [
                r for r in results
                if r.get("subject") == subject and r.get("predicate") == predicate
            ]
        except Exception:
            return []
    
    def add_fact(
        self,
        subject: str,
        predicate: str,
        obj: str,
        confidence: float = 1.0
    ) -> str:
        """Add a fact to the knowledge graph."""
        result = self._client.store(
            content=f"{subject} {predicate} {obj}",
            memory_type="semantic",
            metadata={
                "subject": subject,
                "predicate": predicate,
                "object": obj,
                "confidence": confidence
            }
        )
        return result.get("id", "")

File: packages/typescript-src/llamaindex.ts
/**
 * LlamaIndex integration for Muninn memory system.
 *
 * Provides memory classes for LlamaIndex agents to persist
 * conversation history and knowledge across sessions.
 */

import { ChatMemory, ChatMessage } from "llamaindex/memory";
import { MuninnClient } from "./client";

/**
 * LlamaIndex chat memory backed by Muninn.
 *
 * Usage:
 * ```typescript
 * import { MuninnChatMemory } from "muninn-sdk/llamaindex";
 * import { AgentRunner } from "llamaindex/agent";
 *
 * const memory = new MuninnChatMemory({
 *   apiKey: "muninn_xxx",
 *   organizationId: "my-agent"
 * });
 *
 * const agent = AgentRunner.fromLLM({
 *   llm,
 *   memory
 * });
 * ```
 */
export class MuninnChatMemory implements ChatMemory {
  private client: MuninnClient;
  private tokenLimit: number;

  constructor(options: {
    apiKey: string;
    organizationId?: string;
    baseUrl?: string;
    tokenLimit?: number;
  }) {
    this.client = new MuninnClient({
      apiKey: options.apiKey,
      organizationId: options.organizationId || "default",
      baseUrl: options.baseUrl || "https://api.muninn.au"
    });
    this.tokenLimit = options.tokenLimit || 3000;
  }

  async get(input?: string): Promise<ChatMessage[]> {
    const query = input || "";

    // Search for relevant memories
    const memories = await this.client.memories.search({
      query,
      limit: 20
    });

    // Convert to ChatMessage format
    const messages: ChatMessage[] = [];
    for (const memory of memories) {
      const content = memory.content || "";
      const metadata = memory.metadata || {};
      const role = metadata.role || "user";

      messages.push({
        role: role as "user" | "assistant",
        content,
        additionalKwargs: { memory_id: memory.id }
      } as ChatMessage);
    }

    return messages;
  }

  async getAll(): Promise<ChatMessage[]> {
    return this.get("");
  }

  async put(message: ChatMessage): Promise<void> {
    const role = message.role === "user" ? "user" : "assistant";

    await this.client.memories.store({
      content: message.content,
      type: "conversational",
      metadata: {
        role,
        source: "llamaindex"
      }
    });
  }

  async set(messages: ChatMessage[]): Promise<void> {
    for (const message of messages) {
      await this.put(message);
    }
  }

  async reset(): Promise<void> {
    // Intentional no-op - we don't want to delete memories
  }
}

/**
 * Vector-based memory for LlamaIndex using Muninn.
 *
 * Provides semantic search over stored memories.
 * Best for RAG applications where you need to find
 * relevant context from past conversations.
 */
export class MuninnVectorMemory {
  private client: MuninnClient;
  private similarityThreshold: number;

  constructor(options: {
    apiKey: string;
    organizationId?: string;
    baseUrl?: string;
    similarityThreshold?: number;
  }) {
    this.client = new MuninnClient({
      apiKey: options.apiKey,
      organizationId: options.organizationId || "default",
      baseUrl: options.baseUrl || "https://api.muninn.au"
    });
    this.similarityThreshold = options.similarityThreshold || 0.7;
  }

  async add(content: string, metadata?: Record<string, any>): Promise<string> {
    const result = await this.client.memories.store({
      content,
      type: "semantic",
      metadata: metadata || {}
    });
    return result.id || "";
  }

  async retrieve(query: string, limit: number = 10): Promise<Array<{
    content: string;
    score: number;
    metadata: Record<string, any>;
    id: string;
  }>> {
    const results = await this.client.memories.search({
      query,
      limit,
      search_type: "hybrid" // Uses both keyword and vector search
    });

    // Filter by similarity threshold
    return results
      .filter((r: any) => (r.score || 1.0) >= this.similarityThreshold)
      .map((r: any) => ({
        content: r.content || "",
        score: r.score || 1.0,
        metadata: r.metadata || {},
        id: r.id || ""
      }));
  }

  async delete(memoryId: string): Promise<boolean> {
    try {
      await this.client.delete(memoryId);
      return true;
    } catch {
      return false;
    }
  }

  async clear(): Promise<void> {
    // Intentional no-op
  }
}

/**
 * Knowledge graph memory for LlamaIndex.
 *
 * Stores and retrieves facts as entity-relationship triples.
 * Best for agents that need structured knowledge about
 * entities and their relationships.
 */
export class MuninnKnowledgeGraphMemory {
  private client: MuninnClient;

  constructor(options: {
    apiKey: string;
    organizationId?: string;
    baseUrl?: string;
  }) {
    this.client = new MuninnClient({
      apiKey: options.apiKey,
      organizationId: options.organizationId || "default",
      baseUrl: options.baseUrl || "https://api.muninn.au"
    });
  }

  async getEntityFacts(entityName: string, limit: number = 50): Promise<Array<{
    predicate: string;
    object: string;
    confidence: number;
  }>> {
    try {
      const facts = await this.client.facts.getEntityFacts(entityName);
      return facts.slice(0, limit);
    } catch {
      return [];
    }
  }

  async queryRelation(subject: string, predicate: string): Promise<Array<{
    subject: string;
    predicate: string;
    object: string;
    confidence: number;
  }>> {
    try {
      const results = await this.client.facts.search({
        query: `${subject} ${predicate}`,
        limit: 10
      });
      return results.filter(
        (r: any) => r.subject === subject && r.predicate === predicate
      );
    } catch {
      return [];
    }
  }

  async addFact(
    subject: string,
    predicate: string,
    obj: string,
    confidence: number = 1.0
  ): Promise<string> {
    const result = await this.client.memories.store({
      content: `${subject} ${predicate} ${obj}`,
      type: "semantic",
      metadata: {
        subject,
        predicate,
        object: obj,
        confidence
      }
    });
    return result.id || "";
  }
}